repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -110,18 +116,18 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
 
       info(s"Statement $statementId is cancelling")
 
-      val singleThreadExecutor = ExecutionContext.fromExecutorService(
-        Executors.newSingleThreadExecutor())
       Future {
         if (oldStmtState == StatementState.Running) {
-          while (_statements(statementId).state.get() != StatementState.Cancelled) {
+          val currTime = System.currentTimeMillis()
+          while (_statements(statementId).state.get() != StatementState.Cancelled &&
+            System.currentTimeMillis() - currTime <= TimeUnit.SECONDS.toMillis(30)) {
             _sc.foreach(_.cancelJobGroup(statementId.toString))
-            Thread.sleep(100)
+            Thread.sleep(livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TRIGGER_INTERVAL))
           }
         } else if (oldStmtState == StatementState.Available) {
           _statements(statementId).state.set(StatementState.Cancelled)
```

Comment:
There's one possibility that when cancel is issued, the statement execution is already finished. In this cancel `execute` will not change the state of statement, so we should deliberately set it to `cancelled` here.